tests/test_subparser.py

```diff
@@ -0,0 +1,263 @@
+from dataclasses import dataclass, field
+from typing import Union
+from sys import version_info
+
+from pytest import mark, raises
+
+from pydargs import parse
+
+
+@dataclass
+class Action1:
+    a: int
+    b: str = "abc"
+
+
+@dataclass
+class Action2:
+    c: int = 42
+    b: str = "abc"
+    e: str = field(default="def", metadata=dict(positional=True))
+
+
+@dataclass
+class Action3:
+    d: float
+    e: list[str] = field(default_factory=lambda: ["def"])
+
+
+@dataclass
+class Action4:
+    sub_action: Union[Action1, Action2]
+    string4: str = "four"
+
+
+class TestParseAction:
+    @dataclass
+    class Config:
+        action: Union[Action1, Action2]
+        var: int = 12
+        flag: bool = field(default=False, metadata=dict(as_flags=True))
+
+    @mark.parametrize(
+        "help_string",
+        [
+            "usage: prog [-h] [--var VAR] [--flag | --no-flag]",
+            "{Action1,action1,Action2,action2} ...",
+            "action:  {Action1,action1,Action2,action2}",
+        ],
+    )
+    def test_help(self, capsys, help_string: str):
+        with raises(SystemExit):
+            parse(self.Config, ["--help"], prog="prog")  # type: ignore
+        captured = capsys.readouterr()
+        assert help_string in captured.out.replace("\n", "")
+
+    @mark.skipif(version_info < (3, 10), reason="python3.9 prints s slightly different help")
+    @mark.parametrize(
+        "help_string",
+        ["prog Action1 [-h] --a A [--b B]", "options:  -h"],
+    )
+    def test_action_help(self, capsys, help_string: str):
+        with raises(SystemExit):
+            parse(self.Config, ["Action1", "--help"], prog="prog")  # type: ignore
+        captured = capsys.readouterr()
+        assert help_string in captured.out.replace("\n", "")
+        assert "prog Action1 [-h] --a A [--b B]" in captured.out.replace("\n", "")
+
+    def test_is_required(self, capsys):
+        with raises(SystemExit):
+            parse(self.Config, [])
+        captured = capsys.readouterr()
+        assert "the following arguments are required: action" in captured.err
+
+    def test_flag(self, capsys):
+        config = parse(self.Config, ["--flag", "Action2"])
+        assert isinstance(config.action, Action2)
+        assert config.action.c == 42
+        assert config.action.b == "abc"
+        assert config.flag is True
+
+        # Verify that the order matters
+        with raises(SystemExit):
+            parse(self.Config, ["Action2", "--flag"])
+        captured = capsys.readouterr()
+        assert "unrecognized arguments: --flag" in captured.err
```

Comment:
I guess this is an argparse "shortcoming" but does this mean we do not allow "global" flags, i.e. flags that apply to all subcommands, to be listed after picking the subparser?

this could perhaps be something to mention briefly in the readme

Reply:
I don't think it is a shortcoming -- it is a feature. This means that both the global parser as well as the subparser can have the same option (although that doesn't make things easy to understand).

See my comment below for a suggestion.